demo-circuit/src/blaze_csw/constraints/mod.rs

```diff
@@ -0,0 +1,1652 @@
+    ) -> Option<String> {
+
+        let utxo_data = {
+            if csw_prover_data.utxo_data == CswUtxoProverData::get_phantom() {
+                None
+            } else {
+                Some(csw_prover_data.utxo_data)
+            }
+        };
+
+        let ft_data = {
+            if csw_prover_data.ft_data == CswFtProverData::get_phantom(num_commitment_hashes) {
+                None
+            } else {
+                Some(csw_prover_data.ft_data)
+            }
+        };
+
+        let circuit = CeasedSidechainWithdrawalCircuit::new(
+            sidechain_id,
+            constant,
+            csw_prover_data.sys_data.clone(),
+            Some(csw_prover_data.last_wcert),
+            utxo_data,
+            ft_data,
+            num_commitment_hashes,
+            num_custom_fields,
+        ).unwrap();
+
+        let failing_constraint = debug_circuit(circuit.clone()).unwrap();
+
+        if !debug_only {
+            load_g1_committer_key(1 << 17, 1 << 15).unwrap();
+            let ck_g1 = get_g1_committer_key().unwrap();
+
+            let setup_circuit = CeasedSidechainWithdrawalCircuit::get_instance_for_setup(num_commitment_hashes, num_custom_fields, constant.is_some());
+            let params = CoboundaryMarlin::index(ck_g1.as_ref().unwrap(), setup_circuit.clone()).unwrap();
+            let rng = &mut thread_rng();
+
+            let proof = CoboundaryMarlin::prove(
+                &params.0.clone(),
+                ck_g1.as_ref().unwrap(),
+                circuit,
+                true,
+                Some(rng),
+            )
+            .unwrap();
+
+            let current_public_inputs = {
+                if public_inputs.is_none() {
+                    let mut tmp_public_inputs = Vec::new();
+
+                    if constant.is_some() {
+                        tmp_public_inputs.push(constant.unwrap());
+                    }
+
+                    let csw_sys_data_hash =
+                        CeasedSidechainWithdrawalCircuit::compute_csw_sys_data_hash(
+                            &csw_prover_data.sys_data,
+                            sidechain_id,
+                        )
+                        .unwrap();
+
+                    tmp_public_inputs.push(csw_sys_data_hash);
+                    tmp_public_inputs
+                } else {
+                    public_inputs.unwrap()
+                }
+            };
+
+            // Check that the proof gets correctly verified
+            assert!(CoboundaryMarlin::verify(
+                &params.1.clone(),
+                ck_g1.as_ref().unwrap(),
+                current_public_inputs.as_slice(),
+                &proof
+            )
+            .unwrap());
+
+            // Change one public input each time and check that the proof fails
+            for i in 0..current_public_inputs.len() {
+                let mut wrong_public_inputs = current_public_inputs.clone();
+                wrong_public_inputs[i].double_in_place();
```

Comment:
Actually, double_in_place will not generate a wrong public input if the value of the original one is zero, like a sidechain_id

Reply:
Switched to `add_assign()` with commit [00d8a249d159deeace576b91b1b1c1f895428f3f](https://github.com/HorizenOfficial/zendoo-sc-cryptolib/commit/00d8a249d159deeace576b91b1b1c1f895428f3f).